README
TIMEZONE

Timezone aware date math and formats in JavaScript for Node.js and the browser.

    Read This:      http://bigeasy.github.com/timezone
    Source:         http://github.com/bigeasy/timezone
    Documentation:  http://github.com/bigeasy/timezone/wiki

    Build Status:   http://travis-ci.org/bigeasy/timezone
    Issues:         https://github.com/bigeasy/timezone/issues

    License:        The MIT License

PURPOSE

Timezone is a MicroJS library in pure JavaScript with no dependencies that
provides timezone aware date math and date formatting.

Timezone uses the IANA Database to determine the correct wall clock time
anywhere in the world for any time since the dawn of standardized time.

Timezone formats dates with a full implementation of `strftime` formats,
including the GNU `date` extensions.

Timezone represents time in POSIX time and local time using RFC 3999 date
strings.

Timezone is a full featured standards based time library in pure JavaScript for
under 3K minified and gzipped.

HACKING

$ git clone --quiet https://github.com/bigeasy/timezone.git
$ cd timezone && make && npm install && npm test


CHANGELOG
- Version 1.0.24.

 * Upgrade IANA TZ Database for `2019b`. #328.
 * First day of week less than or equal to specific date. #327.

- Version 1.0.23.

Sat Nov 30 20:40:10 UTC 2019

 * Build with Node.js 12 on Travis CI. #324.
 * Upgrade IANA TZ Database for `2019a`. #323.

- Version 1.0.22.

Thu Jan 24 08:32:11 UTC 2019

 * Upgrade IANA TZ Database for `2018i`. #316.

- Version 1.0.21.

Tue Jan 22 14:05:55 UTC 2019

 * Upgrade IANA TZ Database for `2018h`. #314.

- Version 1.0.20.

Wed Oct 31 12:38:57 UTC 2018

 * Upgrade IANA TZ Database for `2018g`. #311.

- Version 1.0.19.

Wed Oct 31 09:06:38 UTC 2018

 * Upgrade IANA TZ Database for `2018f`. #309.

- Version 1.0.18.

Fri Oct  5 07:15:16 UTC 2018

 * Upgrade IANA TZ Database for `2018e`. #302.
 * Allow negative time zone offset in Rules column. #307.

- Version 1.0.17.

Thu Oct  4 09:06:00 UTC 2018

 * Use shorthand URL in `package.json`. #305.
 * Upgrade IANA TZ Database for `2018d`. #301.

- Version 1.0.16.

Wed Oct  3 16:26:12 UTC 2018

 * Upgrade IANA TZ Database for `2018c`. Closes #300.
 * Tidy commit messages in checklist utility. Closes #303.

- Version 1.0.15.

Wed Oct  3 09:20:37 UTC 2018

 * Remove dimensions from `README.md` image. #299.
 * Upgrade IANA TZ Database for `2018b`. #297.

- Version 1.0.14.

Tue Oct  2 06:39:47 UTC 2018

 * Build with Node.js 10 on Travis CI. #296.
 * Remove `zdump.patch`. #295.
 * Upgrade IANA TZ Database for `2018a`. #293.

- Version 1.0.13.

Sun Nov 19 10:01:52 UTC 2017

 * Upgrade IANA TZ Database for `2017c`. #290.

- Version 1.0.12.

Sun Nov 19 09:10:40 UTC 2017

 * Upgrade IANA TZ Database for `2017b`. #288.

- Version 1.0.11.

Sun Nov 19 08:37:40 UTC 2017

 * Upgrade IANA TZ Database for `2017a`. #286.

- Version 1.0.10.

Sun Nov 19 03:55:44 UTC 2017

 * Upgrade IANA TZ Database for `2016j`. #284.

- Version 1.0.9.

Fri Nov 17 15:53:07 UTC 2017

 * Upgrade IANA TZ Database for `2016i`. #282.

- Version 1.0.8.

Fri Nov 17 14:54:24 UTC 2017

 * Upgrade IANA TZ Database for `2016h`. #280.

- Version 1.0.7.

Fri Nov 17 07:50:56 UTC 2017

 * Upgrade IANA TZ Database to `2016g`. #278.
 * Do not change input date array. #276.

- Version 1.0.6.

Wed Jul 13 16:22:50 UTC 2016

 * Upgrade IANA TZ Database to `2016f`. #275.

- Version 1.0.5.

Wed Jun 15 16:30:23 UTC 2016

 * Upgrade IANA TZ Database to `2016e`. #272.

- Version 1.0.4.

Thu Jun  9 02:36:10 UTC 2016

 * Upgrade IANA TZ Database to `2016d`. #266.

- Version 1.0.3.

Thu Jun  9 02:11:50 UTC 2016

 * Upgrade IANA TZ Database to `2016c`. #269.

- Version 1.0.2.

Thu Jun  9 01:26:34 UTC 2016

 * Upgrade IANA TZ Database to `2016b`. #267.

- Version 1.0.1.

Tue Feb 16 16:13:47 UTC 2016

 * Upgrade IANA TZ Database to `2016a`. #264.

- Version 1.0.0.

Mon Feb 15 20:58:45 UTC 2016

 * Decide that the API is finished.

- Version 0.0.48.

Sun Nov 29 20:29:46 UTC 2015

 * Test with Node.js 5.1 on Travis CI. f5d44667b81356e3f8ea43e2cd682add7d8b39b2.
 * Test with Node.js 4.2 on Travis CI. c56d3e2fc43eb1d2475187de103de7aa4f4042b0.
 * Use require for resource indexes. #259.
 * Continued test tidy. #244.

- Version 0.0.47.

Thu Oct  8 00:03:03 UTC 2015

 * Upgrade IANA TZ Database to `2015g`. #257.
 * More testing tidy.

- Version 0.0.46.

Tue Sep 29 02:29:26 UTC 2015

 * Upgrade TZ Database to `tz2015f`. #255.

- Version 0.0.45.

Sun Sep 27 04:26:41 UTC 2015

 * Upgrade TZ Database to `tz2015e`. #253.

- Version 0.0.44.

Sat Sep 26 04:58:12 UTC 2015

 * Upgrade TZ Database to `tz2015d`. #248.

- Version 0.0.43.

Thu Sep 24 04:40:30 UTC 2015

 * Publish over damaged version 0.0.42 in NPM.

- Version 0.0.42.

Tue Sep 22 06:26:55 UTC 2015

 * Upgrade TZ Database to `tz2015c`. #234.

- Version 0.0.41.

Fri Apr 17 07:45:24 UTC 2015

 * Upgrade TZ Database to `tz2015b`. #232.

- Version 0.0.40.

Fri Feb  6 14:40:10 UTC 2015

 * Upgrade TZ Database to `tz2015a`. #230.

- Version 0.0.39.

Fri Feb  6 13:23:08 UTC 2015

 * Upgrade TZ Database to `tz2014j`. #227.

- Version 0.0.38.

Mon Nov  3 13:40:31 UTC 2014

 * Cannot find module `timezone/loaded`. #223.

- Version 0.0.37.

Thu Oct 30 23:24:49 UTC 2014

 * Upgrade TZ Database to `tz2014i`. #221.

- Version 0.0.36.

Thu Oct 30 22:31:20 UTC 2014

 * Upgrade TZ Database to `tz2014h`. #218.

- Version 0.0.35.

Sun Oct 26 03:17:59 UTC 2014

 * Upgrade TZ Database to `tz2014g`. #217.

- Version 0.0.34.

Thu Aug  7 01:55:45 UTC 2014

 * Upgrade TZ Database to `tz2014f`. #215.

- Version 0.0.33.

Fri Jun 27 16:21:08 UTC 2014

 * Upgrade TZ Database to `tz2014e`. #213.

- Version 0.0.32.

Mon Jun  2 06:34:24 UTC 2014

 * Upgrade TZ Database to `tz2014d`. #211.

- Version 0.0.31.

Sun Jun  1 21:14:47 UTC 2014

 * Upgrade TZ database to `2014b`. #209.
 * Upgrade Proof to 0.0.45. #208.

- Version 0.0.30.

Fri Mar 28 02:32:38 UTC 2014

 * Upgrade TZ database to `2014b`. #206.
 * Upgrade Proof to 0.0.44. #205.

- Version 0.0.29.

Sun Mar 16 04:47:13 UTC 2014

 * Upgrade to 2014a. #202.

- Version 0.0.28.

Wed Feb 12 18:53:25 UTC 2014

 * Upgrade TZ library to `2013i`. #200.

- Version 0.0.27.

Wed Feb 12 16:49:07 UTC 2014

 * Upgrade TZ database to `2013h`. #198.
 * Add punctuation to checklist. #197.

- Version 0.0.26.

Wed Feb 12 03:44:56 UTC 2014

 * Upgrade TZ database to `2013g`. #195.

- Version 0.0.25.

Tue Feb 11 00:08:41 UTC 2014

 * Upgrade TZ database to `tz2013f`. #193.
 * Add missing pushes to checklist. #192.
 * Update checklist with one liners. #191.

- Version 0.0.24.

Sat Feb  8 08:44:49 UTC 2014

 * Upgrade TZ database to `tz2013e`. #189.
 * Remove "url" property from `package.json`. #188.
 * Upgrade Proof to 0.0.42. #187.
 * Fix `homepage` in `package.json`. #186.
 * Add `license` property to `package.json`. #185.
 * Remove Node 0.6 target from Travis CI. #184.
 * Add repository type to `package.json`. #183.
 * Tidy `package.json`. #182.
 * Create a checklist utility. #181.
 * Add `bugs` property to `package.json`. #179.

- Version 0.0.23.

Fri Aug  2 00:17:59 UTC 2013


rfc822.js
// A full RFC 2822 date parser for use with **Timezone**.
//
// This function will return a date array that you can pass to the **Timezone**
// function as the date argument. This is an example of extending **Timezone**
// by building upon it, through functional composition.
//
// RFC 2822 specifies how to parse an RFC 822 date. This is an implementation of
// that specification.
//
// Beware of date strings that kind of look like RFC 822 dates. The similarity
// in appearance between RFC 822 and the `%c` date format specifier in
// `strftime` has led programmers to use the `%c` format for RFC 822. This is
// incorrect.
//
// The format is subtly different from RFC 822 for US English. RFC 822 has a
// comma after the day of week name, `%c` does not. We do not forgive this,
// because it gets worse.
//
// The differences between `%c` and RFC 822 become far more apparent when you
// change locales, replacing the English day of week and month name
// abbreviations with the abbreviations in set locale.
//
// Localization is a general problem. Developers will create a format string for
// an RFC 822 date using day of week and month names that their [date formatting
// library will then localize](//
// http://rel.me/2008/07/22/date-format-rfc82285010361123asctimeiso8601unicode35tr35-6/). 
//
// While `So 01 Mär 1992 00:00:00 UTC` is not uncommon, it is still not an RFC
// 822 date. RFC 822 dates are not localized. They are always English according
// to the RFC. Therefore, our RFC 2822 date parser expects English and does not
// support localization.
//
// If enough examples of accidentally localized RFC 822 dates are encountered in
// the wild, we can change that. Let me know. 

/*
var __slice = [].slice
function die () {
  console.log.apply(console, __slice.call(arguments, 0));
  return process.exit(1);
};

function say () { return console.log.apply(console, __slice.call(arguments, 0)) }
*/
!function (definition) {
  if (typeof module == "object" && module.exports) module.exports = definition();
  else if (typeof define == "function") define(definition);
  else this.tz = definition();
} (function () {
  // Our parser is case insensitive for the sake of forgiveness, even though the
  // specification is pretty clear about case.
  var DAY = "sun|mon|tue|wed|thu|fri|sat".split("|")
    , MONTH = "jan|feb|mar|apr|may|jun|jul|aug|sep|oct|nov|dec".split("|")
    // RFC 822 specified single character military time zone abbreviations, but
    // got them wrong, reversed from the actual time zones used by NATO. Thus,
    // RFC 2822 says that they can't be trusted. The only abbreviations that can
    // be trusted are the synonyms for UTC, and the abbreviations for the four
    // time zones of the contiguous United States. These were laid out in RFC
    // 822.
    //
    // See [RFC 1123 5.2.14 Date and Time
    // Specification](http://www.ietf.org/rfc/rfc1123.txt).
    //
    // Otherwise, you cannot know the time zone offset from a date string found
    // in the wild, only in an RFC 822 date and only if is one of the following.
    , ABBREV =
      { Z: "+0000"
      , UT: "+0000"
      , GMT: "+0000"
      , UTC: "+0000"
      , EDT: "-0400"
      , EST: "-0500"
      , CDT: "-0500"
      , CST: "-0600"
      , MDT: "-0600"
      , MST: "-0700"
      , PDT: "-0700"
      , PST: "-0800"
      }
    //
    , strip
    , $
    ;

  // The RFC 822 function is our sole export.
  return rfc822;

  // In the IANA Timezone Database, Arthur David Olson has this to say about
  // translating time zone abbreviations to time zone offsets.

  // > The UNIX Version 7 "timezone" function is not present in this package; it's
  // > impossible to reliably map timezone's arguments (a "minutes west of GMT"
  // > value and a "daylight saving time in effect" flag) to a time zone
  // > abbreviation, and we refuse to guess.

  // So, please, no pull requests with more time zone abbreviation mappings.
  // What might be right for you, may not be right for some.
  //
  // RFC 2822 says that you can interpret military zones if you have out-of-band
  // information confirming whether they are NATO or RFC 822. The second
  // parameter to our parsing function is this optional out-of-band data.
  //
  //     var eq = require("assert").equal
  //       , tz = require("timezone")
  //       , moonwalk = tz("1969-07-21 02:56")
  //       , military = { nato: {}, rfc822: {} }
  //       , index
  //       ;
  //
  //     index = 0;
  //     "ABCDEFGHIKLM".replace(/./g, function (ch) {
  //       var offset = ("0" + (++index) + "00").slice(-4);
  //       military.nato[ch] = "+" + offset;
  //       military.rfc822[ch] = "-" + offset;
  //     });
  //
  //     index = 0;
  //     "NOPQRSTUVWXY".replace(/./g, function (ch) {
  //       var offset = ("0" + (++index) + "00").slice(-4);
  //       military.nato[ch] = "-" + offset;
  //       military.rfc822[ch] = "+" + offset;
  //     });
  //
  //     eq( tz(rfc822("Sun, 20 Jul 1969 10:56:00 E", military.rfc822)), moonwalk );
  //     eq( tz(rfc822("Sun, 20 Jul 1969 10:56:00 R", military.nato)), moonwalk );
  //
  // If you have any additional out-of-band information on time zone
  // abbreviations for your application, provide that information in the map of
  // additional abbreviations.

  //
  function rfc822(date, abbrev) {
    abbrev = abbrev || {};

    // The full RFC 2822 specification includes support for features of legacy
    // RFC 822 dates that I've never see in the wild. Specifically the ability
    // to place comments anywhere in a date.
    //
    // > `Sun, 1 Mar (Spring is coming soon!) 92 00:00:00 GMT`
    //
    // Comments are enclosed in parenthesis. Comments can contain nested
    // comments making it impossible to match them with a regular expression, as
    // any discussion of a regular expression to match an RFC 822 email address
    // will tell you.
    //
    // We strip the comments. Our approach to RFC 822 comments is informed by a
    // long peek at the workings of
    // [Mail::RFC822::Address](http://www.ex-parrot.com/pdw/Mail-RFC822-Address.html).
    do { strip = date } while ((date = strip.replace(/\((?:[^()\\]|\\.)*\)/, " ")) != strip);

    // To parse the named timezones we'll simply match 1 or more consecutive
    // letters, a through z. If those letters form on one of the time zone
    // abbreviations we know about, we use the time zone offset associated with
    // the abbreviation. Otherwise, we use `"-0000"`.

    //
    $ = /^\s*(?:(\w{3}),\s+)?(\d{1,2})\s+(\w{3})\s+(\d{2,4})\s+(\d{2}):(\d{2})(?::(\d{2}))?\s*(?:([-+]?\d{4})|([a-z]{1,5}))?\s*$/i.exec(date)
    if (!$) throw new Error("invalid rfc822 date");

    $[3] = MONTH.indexOf($[3].toLowerCase());
    if (!~$[3]) throw new Error("invalid month");

    $[0] = +($[4]);
    $[4] = $[2];
    $[2] = $[0];

    // RFC 2822 specifies how to interpret a 2 digit year or 3 digit year exactly.
    if ($[2] < 50) $[2] += 2000;
    else if ($[2] < 1000) $[2] += 1900;

    date = $.slice(2, 8);

    // If it comes to bite you that someone has specified the wrong day of week
    // for a particular date, you probably cannot trust those dates, but if you
    // must parse them, strip the day of the week from the RFC 822 date string
    // before calling this function to skip this assertion.

    //
    if ($[1]) {
      $[1] = DAY.indexOf($[1].toLowerCase());
      if (!~$[1])
        throw new Error("invalid day of week");
      if (new Date(Date.UTC.apply(Date.UTC, date)).getUTCDay() != $[1])
        throw new Error("incorrect day of week for date");
    }

    date.push(0);

    // If we cannot be certain of a time zone abbreviation, we default to
    // "-0000" which indicates an uncertain UTC, according to RFC 2822. Of
    // course, that bit of knowledge never leaves the function, but we'll just
    // do as RFC 2822 says.

    //
    offset = $[8] || ABBREV[$[9].toUpperCase()] || abbrev[$[9].toUpperCase()] || "-0000"
    if (offset == "GMT") date.push(1);
    else if (offset) date.push(offset < 0 ? -1 : 1, Math.abs(Math.floor(offset / 100)), offset % 100);

    ++date[1];

    return date;
  }
});
